File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
from collections import defaultdict
def solution(want, number, discount):
    answer = 0
    
    dict = {}
    i=0
    for w in want:
        dict[w] = number[i]
        i+=1
    

    for i in range(len(discount)-9):
        tmp = dict.copy()
        for j in range(i,i+10):
            if discount[j] in tmp and tmp[discount[j]] !=0:
                tmp[discount[j]]-=1
            else:
                break
        if sum(tmp.values()) == 0:
            answer +=1

    return answer

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;
class Solution {
    static HashMap<String, Integer> map;
    static HashMap<String, Integer> wmap;
    static List<String> keys;
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        
        map = new HashMap<>();
        wmap = new HashMap<>();
        for(int i=0; i<10;i++){
            map.put(discount[i], map.getOrDefault(discount[i],0)+1 );
        }
        
        for(int i=0; i<want.length;i++){
            wmap.put(want[i], number[i]);
        }
        
        keys = Arrays.asList(want);
        if(solve())answer++;
        for(int i=10; i<discount.length;i++){
            map.put(discount[i-10], map.get(discount[i-10])-1);
            map.put(discount[i], map.getOrDefault(discount[i],0)+1);
            if(solve())answer++;
        }
        
        
        return answer;
    }
    public static boolean solve(){
        for(String key: keys){
            if(map.getOrDefault(key,0)<wmap.get(key))return false;
        }
            return true;
    }
}